File: android/app/src/main/kotlin/com/baseapp/android/view/games/gomoku/GomokuGameView.kt
package com.baseapp.android.view.games.gomoku

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.baseapp.android.view.theme.LocalUiTheme
import kotlinx.coroutines.delay

/**
 * Wooden goban palette mirroring the web GomokuPage styles. The board keeps
 * its own warm wood look on both light and dark app themes (why: the web
 * playing surface is hardcoded wood too — it is part of the game's identity).
 */
private object GomokuColors {
    val wood = Color(0xFFDEB26A)
    val woodEdge = Color(0xFFB98A45)
    val gridLine = Color(0xBF3B2812) // rgba(59,40,18,0.75)
    val stoneShadow = Color(0x731E1004) // rgba(30,16,4,0.45)
    val blackHighlight = Color(0xFF5C5C60)
    val blackBody = Color(0xFF0A0A0C)
    val whiteHighlight = Color(0xFFFFFFFF)
    val whiteBody = Color(0xFFC9C2AE)
    val lastMove = Color(0xFFE05C3A)
    val winGlow = Color(0xFFFFD97A)
}

/** Milliseconds the bot "thinks" before replying, like the web page. */
private const val BOT_THINK_DELAY_MS = 300L

/**
 * Home surface for the `gomoku` pack — the native twin of the web GomokuPage.
 * Rendering and touch input only: all rules live in [GomokuEngine] so the
 * game stays JVM-testable and in lockstep with the web engine. No network,
 * no stores.
 *
 * Vs-bot only (the web 2P hotseat mode is dropped on mobile — tap-to-place
 * with the same three bot levels covers the core game). The human plays
 * black and always moves first.
 */
@Composable
fun GomokuGameView() {
    val theme = LocalUiTheme.current
    val engine = remember { GomokuEngine() }
    var level by remember { mutableStateOf(GomokuEngine.BotLevel.MEDIUM) }
    // Bumped after every board mutation so the Canvas and status recompose.
    var moveCount by remember { mutableIntStateOf(0) }
    var botThinking by remember { mutableStateOf(false) }

    // Bot reply: deferred so the player's stone paints first, like the web.
    LaunchedEffect(botThinking) {
        if (botThinking) {
            delay(BOT_THINK_DELAY_MS)
            engine.playBotMove()
            botThinking = false
            moveCount = engine.moveCount
        }
    }

    fun newGame() {
        engine.newGame()
        botThinking = false
        moveCount = engine.moveCount
    }

    val statusText = when {
        engine.winner == GomokuEngine.Stone.BLACK -> "YOU WIN — FIVE IN A ROW"
        engine.winner == GomokuEngine.Stone.WHITE -> "BOT WINS — FIVE IN A ROW"
        engine.isDraw -> "DRAW — BOARD FULL"
        botThinking -> "BOT THINKING…"
        else -> "YOUR MOVE (BLACK)"
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(theme.colors.appBg)
            .statusBarsPadding()
            .padding(theme.spacing.lg),
        verticalArrangement = Arrangement.spacedBy(theme.spacing.md),
    ) {
        Text(
            text = "GOMOKUBOT",
            color = theme.colors.textPrimary,
            fontSize = theme.typography.sizes.xl,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Monospace,
        )

        Row(horizontalArrangement = Arrangement.spacedBy(theme.spacing.sm)) {
            RetroButton(label = "⟳ NEW GAME", onClick = ::newGame)
            RetroButton(
                label = "↩ UNDO",
                onClick = {
                    if (engine.moveCount > 0 && !botThinking) {
                        engine.undoPair()
                        moveCount = engine.moveCount
                    }
                },
            )
        }

        Box(
            modifier = Modifier
                .weight(1f)
                .fillMaxWidth(),
            contentAlignment = Alignment.Center,
        ) {
            GomokuBoardCanvas(
                engine = engine,
                moveCount = moveCount,
                onTapCell = { cell ->
                    if (!botThinking &&
                        engine.turn == GomokuEngine.Stone.BLACK &&
                        engine.place(cell)
                    ) {
                        moveCount = engine.moveCount
                        if (!engine.isOver) {
                            botThinking = true
                        }
                    }
                },
                modifier = Modifier
                    .aspectRatio(1f)
                    .clip(RoundedCornerShape(theme.radius.md))
                    .border(2.dp, GomokuColors.woodEdge, RoundedCornerShape(theme.radius.md)),
            )
        }

        Row(horizontalArrangement = Arrangement.spacedBy(theme.spacing.sm)) {
            GomokuEngine.BotLevel.entries.forEach { option ->
                RetroButton(
                    label = option.name,
                    isLit = option == level,
                    onClick = {
                        level = option
                        engine.level = option
                    },
                )
            }
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            StatusText(text = statusText)
            StatusText(text = "MOVE $moveCount")
        }
    }
}

/**
 * The goban: draws wood, grid lines, star points, stones with a subtle drop
 * shadow, the last-move marker, and the winning-line glow. Taps are mapped
 * back to the nearest intersection.
 */
@Composable
private fun GomokuBoardCanvas(
    engine: GomokuEngine,
    moveCount: Int,
    onTapCell: (Int) -> Unit,
    modifier: Modifier = Modifier,
) {
    // Star-point (hoshi) rows/columns of a 15x15 goban.
    val starLines = remember { listOf(3, 7, 11) }

    Canvas(
        modifier = modifier.pointerInput(engine) {
            detectTapGestures { position ->
                val cellSize = size.width / GomokuEngine.BOARD_SIZE.toFloat()
                val col = (position.x / cellSize).toInt()
                val row = (position.y / cellSize).toInt()
                if (row in 0 until GomokuEngine.BOARD_SIZE &&
                    col in 0 until GomokuEngine.BOARD_SIZE
                ) {
                    onTapCell(GomokuEngine.cellAt(row, col))
                }
            }
        },
    ) {
        // Reading moveCount invalidates the canvas after every placement; the
        // engine itself is not snapshot state.
        @Suppress("UNUSED_EXPRESSION")
        moveCount

        val cellSize = size.width / GomokuEngine.BOARD_SIZE
        val inset = cellSize / 2f
        val far = size.width - inset

        drawRect(color = GomokuColors.wood)

        // Grid lines: intersections sit at cell centers, like the web board.
        for (index in 0 until GomokuEngine.BOARD_SIZE) {
            val offset = (index + 0.5f) * cellSize
            drawLine(
                color = GomokuColors.gridLine,
                start = Offset(inset, offset),
                end = Offset(far, offset),
                strokeWidth = 1.5f,
            )
            drawLine(
                color = GomokuColors.gridLine,
                start = Offset(offset, inset),
                end = Offset(offset, far),
                strokeWidth = 1.5f,
            )
        }

        // Star points (hoshi).
        for (row in starLines) {
            for (col in starLines) {
                if (engine.board[GomokuEngine.cellAt(row, col)] == null) {
                    drawCircle(
                        color = GomokuColors.gridLine,
                        radius = cellSize * 0.09f,
                        center = intersectionCenter(row, col, cellSize),
                    )
                }
            }
        }

        // Stones, then the win glow / last-move marker.
        val winCells = engine.winLine?.toSet() ?: emptySet()
        val stoneRadius = cellSize * 0.42f
        for (cell in 0 until GomokuEngine.CELL_COUNT) {
            val stone = engine.board[cell] ?: continue
            val center = intersectionCenter(GomokuEngine.rowOf(cell), GomokuEngine.colOf(cell), cellSize)

            drawCircle(
                color = GomokuColors.stoneShadow,
                radius = stoneRadius,
                center = center + Offset(cellSize * 0.04f, cellSize * 0.06f),
            )
            drawCircle(
                brush = Brush.radialGradient(
                    colors = if (stone == GomokuEngine.Stone.BLACK) {
                        listOf(GomokuColors.blackHighlight, GomokuColors.blackBody)
                    } else {
                        listOf(GomokuColors.whiteHighlight, GomokuColors.whiteBody)
                    },
                    center = center + Offset(-stoneRadius * 0.35f, -stoneRadius * 0.4f),
                    radius = stoneRadius * 1.6f,
                ),
                radius = stoneRadius,
                center = center,
            )

            if (cell in winCells) {
                drawCircle(
                    color = GomokuColors.winGlow,
                    radius = stoneRadius + cellSize * 0.05f,
                    center = center,
                    style = Stroke(width = cellSize * 0.08f),
                )
            } else if (cell == engine.lastMove) {
                drawCircle(
                    color = GomokuColors.lastMove,
                    radius = cellSize * 0.1f,
                    center = center,
                )
            }
        }
    }
}

private fun intersectionCenter(row: Int, col: Int, cellSize: Float): Offset =
    Offset((col + 0.5f) * cellSize, (row + 0.5f) * cellSize)

/** Chunky monospace control button echoing the web console's toolbar. */
@Composable
private fun RetroButton(label: String, isLit: Boolean = false, onClick: () -> Unit) {
    val theme = LocalUiTheme.current
    val shape = RoundedCornerShape(theme.radius.sm)
    Box(
        modifier = Modifier
            .clip(shape)
            .background(if (isLit) theme.colors.accent else theme.colors.surfaceAlt)
            .border(1.dp, theme.colors.border, shape)
            .clickable(onClick = onClick)
            .padding(horizontal = 10.dp, vertical = 8.dp),
    ) {
        Text(
            text = label,
            color = if (isLit) theme.colors.accentText else theme.colors.textPrimary,
            fontSize = theme.typography.sizes.xs,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Monospace,
        )
    }
}

@Composable
private fun StatusText(text: String) {
    val theme = LocalUiTheme.current
    Text(
        text = text,
        color = theme.colors.textSecondary,
        fontSize = theme.typography.sizes.xs,
        fontFamily = FontFamily.Monospace,
    )
}

File: ios/App/View/Games/Gomoku/GomokuGameView.swift
import SwiftUI

/// Home surface for the `gomoku` pack — the native twin of the web
/// `GomokuPage`. Purely client-side: gomoku projects have no backend, so this
/// view must never touch stores, components, or the network.
///
/// Vs-bot only (the web 2P hotseat mode is dropped on mobile — tap-to-place
/// with the same three bot levels covers the core game). The human plays
/// black and always moves first; all rules live in `GomokuEngine`.
///
/// A single `Canvas` renders the whole goban (wood, grid, star points,
/// stones, markers) like the web board, so a tap only re-renders once via the
/// `moveCount` state bump.
struct GomokuGameView: View {
  @Environment(\.uiThemeTokens) private var theme

  /// Reference-typed game state; SwiftUI re-renders via `moveCount`.
  @State private var engine = GomokuEngine()
  @State private var level: GomokuBotLevel = .medium
  /// Bumped after every board mutation so the Canvas invalidates.
  @State private var moveCount = 0
  @State private var botThinking = false

  var body: some View {
    VStack(spacing: theme.spacing.lg) {
      Text("GOMOKUBOT")
        .font(.system(size: theme.typography.sizes.xl, weight: .bold, design: .monospaced))
        .foregroundStyle(theme.colors.textPrimary)
        .padding(.top, theme.spacing.lg)

      GomokuBoardView(engine: engine, moveCount: moveCount, onTapCell: handleTap)
        .aspectRatio(1, contentMode: .fit)
        .padding(.horizontal, theme.spacing.md)

      Text(statusText)
        .font(.system(size: theme.typography.sizes.sm, weight: .bold, design: .monospaced))
        .foregroundStyle(theme.colors.textPrimary)

      controls

      Text("Tap an intersection to place a stone · Five in a row wins")
        .font(.system(size: theme.typography.sizes.xs, design: .monospaced))
        .foregroundStyle(theme.colors.textSecondary)

      Spacer(minLength: 0)
    }
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(theme.colors.appBg)
    .onChange(of: level) { _, newValue in
      engine.level = newValue
    }
  }

  private var statusText: String {
    if let winner = engine.winner {
      return winner == .black ? "YOU WIN — FIVE IN A ROW" : "BOT WINS — FIVE IN A ROW"
    }
    if engine.isDraw {
      return "DRAW — BOARD FULL"
    }
    return botThinking ? "BOT THINKING…" : "YOUR MOVE (BLACK)"
  }

  private var controls: some View {
    VStack(spacing: theme.spacing.md) {
      HStack(spacing: theme.spacing.md) {
        Button {
          engine.newGame()
          botThinking = false
          moveCount = engine.moves.count
        } label: {
          Label("New Game", systemImage: "arrow.counterclockwise")
        }
        .buttonStyle(GomokuChunkyButtonStyle())

        Button {
          engine.undoPair()
          moveCount = engine.moves.count
        } label: {
          Label("Undo", systemImage: "arrow.uturn.backward")
        }
        .buttonStyle(GomokuChunkyButtonStyle())
        .disabled(engine.moves.isEmpty || botThinking)
      }

      Picker("Bot Level", selection: $level) {
        ForEach(GomokuBotLevel.allCases) { option in
          Text(option.rawValue.capitalized).tag(option)
        }
      }
      .pickerStyle(.segmented)
      .padding(.horizontal, theme.spacing.md)
    }
  }

  private func handleTap(cell: Int) {
    guard !botThinking, engine.turn == .black, engine.place(cell: cell) else { return }
    moveCount = engine.moves.count
    guard !engine.isOver else { return }

    // Bot reply: deferred so the player's stone paints first, like the web.
    botThinking = true
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
      engine.playBotMove()
      botThinking = false
      moveCount = engine.moves.count
    }
  }
}

/// Chunky retro button consistent with the web page's toolbar buttons.
private struct GomokuChunkyButtonStyle: ButtonStyle {
  @Environment(\.uiThemeTokens) private var theme

  func makeBody(configuration: Configuration) -> some View {
    configuration.label
      .font(.system(size: theme.typography.sizes.sm, weight: .bold, design: .monospaced))
      .foregroundStyle(theme.colors.textPrimary)
      .padding(.horizontal, 14)
      .padding(.vertical, 9)
      .background(
        RoundedRectangle(cornerRadius: theme.radius.sm, style: .continuous)
          .fill(configuration.isPressed ? theme.colors.hover : theme.colors.surfaceAlt)
          .overlay(
            RoundedRectangle(cornerRadius: theme.radius.sm, style: .continuous)
              .stroke(theme.colors.border, lineWidth: 1)
          )
      )
  }
}

/// The goban: draws wood, grid lines, star points, stones, the last-move
/// marker, and the winning-line glow with the web board's palette. Taps are
/// mapped back to the nearest intersection.
private struct GomokuBoardView: View {
  let engine: GomokuEngine
  /// Only read to invalidate the Canvas when the parent bumps it.
  let moveCount: Int
  let onTapCell: (Int) -> Void

  /// Star-point (hoshi) rows/columns of a 15x15 goban.
  private static let starLines = [3, 7, 11]

  var body: some View {
    GeometryReader { proxy in
      let side = min(proxy.size.width, proxy.size.height)
      let cellSize = side / CGFloat(GomokuEngine.boardSize)

      Canvas { context, _ in
        _ = moveCount
        draw(context: context, side: side, cellSize: cellSize)
      }
      .gesture(
        DragGesture(minimumDistance: 0)
          .onEnded { value in
            let col = Int(value.location.x / cellSize)
            let row = Int(value.location.y / cellSize)
            guard (0..<GomokuEngine.boardSize).contains(row),
                  (0..<GomokuEngine.boardSize).contains(col)
            else { return }
            onTapCell(GomokuEngine.cellAt(row: row, col: col))
          }
      )
    }
    .background(Color(hex: "#deb26a"))
    .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: 8, style: .continuous)
        .stroke(Color(hex: "#b98a45"), lineWidth: 2)
    )
  }

  private func center(of cell: Int, cellSize: CGFloat) -> CGPoint {
    CGPoint(
      x: (CGFloat(GomokuEngine.colOf(cell)) + 0.5) * cellSize,
      y: (CGFloat(GomokuEngine.rowOf(cell)) + 0.5) * cellSize
    )
  }

  private func draw(context: GraphicsContext, side: CGFloat, cellSize: CGFloat) {
    let gridColor = Color(hex: "#3b2812").opacity(0.75)
    let inset = cellSize / 2
    let far = side - inset

    // Grid lines: intersections sit at cell centers, like the web board.
    var grid = Path()
    for index in 0..<GomokuEngine.boardSize {
      let offset = (CGFloat(index) + 0.5) * cellSize
      grid.move(to: CGPoint(x: inset, y: offset))
      grid.addLine(to: CGPoint(x: far, y: offset))
      grid.move(to: CGPoint(x: offset, y: inset))
      grid.addLine(to: CGPoint(x: offset, y: far))
    }
    context.stroke(grid, with: .color(gridColor), lineWidth: 1)

    // Star points (hoshi).
    for row in Self.starLines {
      for col in Self.starLines where engine.board[GomokuEngine.cellAt(row: row, col: col)] == nil {
        let point = center(of: GomokuEngine.cellAt(row: row, col: col), cellSize: cellSize)
        let radius = cellSize * 0.09
        context.fill(
          Path(ellipseIn: CGRect(
            x: point.x - radius, y: point.y - radius, width: radius * 2, height: radius * 2
          )),
          with: .color(gridColor)
        )
      }
    }

    // Stones with a subtle drop shadow, then the win glow / last-move marker.
    let winCells = Set(engine.winLine ?? [])
    let stoneRadius = cellSize * 0.42
    for cell in 0..<GomokuEngine.cellCount {
      guard let stone = engine.board[cell] else { continue }
      let point = center(of: cell, cellSize: cellSize)
      let rect = CGRect(
        x: point.x - stoneRadius, y: point.y - stoneRadius,
        width: stoneRadius * 2, height: stoneRadius * 2
      )

      context.fill(
        Path(ellipseIn: rect.offsetBy(dx: cellSize * 0.04, dy: cellSize * 0.06)),
        with: .color(Color(hex: "#1e1004").opacity(0.45))
      )
      context.fill(
        Path(ellipseIn: rect),
        with: .radialGradient(
          Gradient(colors: stone == .black
            ? [Color(hex: "#5c5c60"), Color(hex: "#0a0a0c")]
            : [Color.white, Color(hex: "#c9c2ae")]),
          center: CGPoint(x: point.x - stoneRadius * 0.35, y: point.y - stoneRadius * 0.4),
          startRadius: 0,
          endRadius: stoneRadius * 1.6
        )
      )

      if winCells.contains(cell) {
        context.stroke(
          Path(ellipseIn: rect.insetBy(dx: -cellSize * 0.05, dy: -cellSize * 0.05)),
          with: .color(Color(hex: "#ffd97a")),
          lineWidth: cellSize * 0.08
        )
      } else if cell == engine.lastMove {
        let radius = cellSize * 0.1
        context.fill(
          Path(ellipseIn: CGRect(
            x: point.x - radius, y: point.y - radius, width: radius * 2, height: radius * 2
          )),
          with: .color(Color(hex: "#e05c3a"))
        )
      }
    }
  }
}
